File: 1019-squares-of-a-sorted-array/squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> result(nums.size());
        int l = 0, r = nums.size() - 1, i = nums.size() - 1;
        while (l <= r) {
            if (abs(nums[l]) < abs(nums[r]))
                result[i--] = nums[r] * nums[r--]; 
            else
                result[i--] = nums[l] * nums[l++]; 
        }
        return result;
    }
};

File: 1019-squares-of-a-sorted-array/squares-of-a-sorted-array.go
func sortedSquares(nums []int) []int {
    n := len(nums)
    res := make([]int, n)
    p1, p2, p3 := 0, n-1, n-1

    for p1 <= p2 {
        lsquare := nums[p1] * nums[p1]
        rsquare := nums[p2] * nums[p2]

        if lsquare < rsquare {
            res[p3] = rsquare
            p2--
        } else {
            res[p3] = lsquare
            p1++
        }
        p3--
    }

    return res
}